### PR TITLE
Support basic synchronization between external bidder & Openwrap

### PR DESCRIPTION
**Pull request by pm-dattaprasad-mundada**

### CODE REVIEW

File: src_new/owt.js [hunk 1]
```diff
@@ -57,16 +57,35 @@ window.PWT.sfDisplayPMPCreative = function(theDocument, values, priorityArray){
 };
 
 
-window.OWT = {};
+window.OWT = {
```
Comment (PubMatic-OpenWrap) [inline src_new/owt.js]: why we have not used PWT name-space ? 
I had given same review-comment earlier as well.

File: src_new/util.js [hunk 2]
```diff
@@ -830,15 +830,28 @@ exports.vLogInfo = function(divID, infoObject){
...
+
+exports.resetExternalBidderStatus = function(divIds) {
+	refThis.forEachOnArray(divIds, function (key, divId) {
+		console.log("resetExternalBidderStatus: ", divId);
+		window.OWT.externalBidderStatuses[divId] = undefined;
+	});
 };
 
-exports.getAllPartnersBidStatuses = function (bidMaps) {
+exports.getAllPartnersBidStatuses = function (bidMaps, divIds) {
```
Comment (PubMatic-OpenWrap) [inline src_new/util.js]: This method can be moved to bidManager as functionality is specific to bids

File: src_new/util.js [hunk 3]
```diff
@@ -830,15 +830,28 @@ exports.vLogInfo = function(divID, infoObject){
 	}
 };
 
-exports.getExternalBidderStatus = function() {
-	return window.OWT.externalBidderStatus;
+exports.getExternalBidderStatus = function(divIds) {
```
Comment (PubMatic-OpenWrap) [inline src_new/util.js]: This method can be moved OWT/PWT name-space as it is specific to a functionality

File: src_new/util.js [hunk 4]
```diff
@@ -830,15 +830,28 @@ exports.vLogInfo = function(divID, infoObject){
...
+	var status = true;
+	refThis.forEachOnArray(divIds, function (key, divId) {
+		status =  window.OWT.externalBidderStatuses[divId]
+							? status && window.OWT.externalBidderStatuses[divId].status
+							: status;
+	});
+	return status;
+};
+
+exports.resetExternalBidderStatus = function(divIds) {
```
Comment (PubMatic-OpenWrap) [inline src_new/util.js]: This method can be moved OWT/PWT name-space as it is specific to a functionality

File: src_new/controllers/gpt.js [hunk 5]
```diff
@@ -529,13 +529,14 @@ exports.processDisplayCalledSlot = processDisplayCalledSlot;
 /* end-test-block */
 
 
-function executeDisplay(timeout, callback) {
-    if (util.getExternalBidderStatus() && util.getAllPartnersBidStatuses(window.PWT.bidMap)) {
-	window.OWT.externalBidderStatus=undefined; //Quick fix to reset flag so that the notification flow happens only once per page load
+function executeDisplay(timeout, divIds, callback) {
```
Comment (PubMatic-OpenWrap) [inline src_new/controllers/gpt.js]: add more test case

File: test/controllers/gpt.spec.js [hunk 6]
```diff
@@ -2018,6 +2018,52 @@ describe("CONTROLLER: GPT", function() {
         });
     });
 
+    describe("#executeDisplay", function() {
+        beforeEach(function(done) {
+            sinon.spy(window, "setTimeout");
+            window.OWT = {
+              externalBidderStatuses: {
```
Comment (PubMatic-OpenWrap) [inline test/controllers/gpt.spec.js]: use actual api calls

File: src_new/controllers/gpt.js [hunk 7]
```diff
@@ -555,15 +556,17 @@ function displayFunctionStatusHandler(oldStatus, theObject, originalFunction, ar
             // eslint-disable-line no-fallthrough
         /* istanbul ignore next */
         case CONSTANTS.SLOT_STATUS.PARTNERS_CALLED:
+            var divIds = Object.keys(refThis.slotsMap);
 
-            if ((typeof window.OWT.externalBidderStatus) === "boolean") {
-               refThis.executeDisplay(CONFIG.getTimeout(), function() {
+            if (window.OWT.externalBidderStatuses[arg[0]]) {
```
Comment (PubMatic-OpenWrap) [inline src_new/controllers/gpt.js]: add type check

File: src_new/owt.js (hunk 1 above)
Comment (PubMatic-OpenWrap) [reply src_new/owt.js]: We will also need to ask IAC to make changes when we move OWT to PWT

File: test/controllers/gpt.spec.js [hunk 8]
```diff
@@ -2031,6 +2077,9 @@ describe("CONTROLLER: GPT", function() {
                 return "originalFunction"
             };
             arg = {};
+            window.OWT = {
```
Comment (PubMatic-OpenWrap) [inline test/controllers/gpt.spec.js]: Need to update test cases, do we need to define window.OWT here ?

File: test/controllers/gpt.spec.js [hunk 9]
```diff
@@ -2726,6 +2779,7 @@ describe("CONTROLLER: GPT", function() {
             GPT.updateSlotsMapFromGoogleSlots.restore();
             GPT.forQualifyingSlotNamesCallAdapters.restore();
             theObject = null;
+            window.OWT = null;
```
Comment (PubMatic-OpenWrap) [inline test/controllers/gpt.spec.js]: check if this can be avoided

File: test/util.spec.js [hunk 10]
```diff
@@ -2105,4 +2105,163 @@ describe('UTIL', function() {
             done();
         });
     });
+
+    describe('#getExternalBidderStatus', function() {
+        beforeEach(function(done) {
+            window.OWT = {
+            	notifyCount: 0,
+            	externalBidderStatuses: {
```
Comment (PubMatic-OpenWrap) [inline test/util.spec.js]: use actual apis

File: test/util.spec.js [hunk 11]
```diff
@@ -2105,4 +2105,163 @@ describe('UTIL', function() {
...
+        it('should return true if external bidder already responded', function(done) {
+            UTIL.getExternalBidderStatus(["Div2"]).should.be.true;
+            done();
+        });
+    });
+
+    describe('#resetExternalBidderStatus', function() {
+        beforeEach(function(done) {
+            window.OWT = {
+            	notifyCount: 0,
```
Comment (PubMatic-OpenWrap) [inline test/util.spec.js]: use actual APIs

File: test/util.spec.js [hunk 12]
```diff
@@ -2105,4 +2105,163 @@ describe('UTIL', function() {
...
+            UTIL.resetExternalBidderStatus(["Div1", "Div2"]);
+            window.OWT.externalBidderStatuses.should.deep.equal({
+              "Div1": undefined,
+              "Div2": undefined,
+            });
+            done();
+        });
+    });
+
+    describe('#getAllPartnersBidStatuses', function() {
```
Comment (PubMatic-OpenWrap) [inline test/util.spec.js]: need to update

**Comment by pm-dattaprasad-mundada** [comment]
1.Currently we are staying with window.OWT namespace, will be ready with window.PWT for stable version. 
2. In test cases we are unable to access window.OWT as its in owt.js which dosen't export anything, so we need to define window.OWT everywhere in testcases.

**Comment by pm-dattaprasad-mundada** [comment]
Now we are creating pull request to Continuous Integration rather than external-bidder-pre-timeout branch.

**Pull request closed without merging**